File: runme/mcgen/local/datan/solsrc/s5lsq.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern double CBFUNC bwfnct(double const P_T *x, integer const P_T *nr,
                            double const P_T *t);
extern double CBFUNC gsfnct(double const P_T *x, integer const P_T *nr,
                            double const P_T *t);

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__3 = 3, c__4 = 4, c__19 = 19, c__21 = 21;
static double c_d0 = 0., c_dp2 = .2, c_dp9 = .9, c_dp17 = .17, c_dp87 = .87,
              c_dmp414 = -.414, c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, iswit, lcapt, list[2], nstep, nws, nx, ny, nr = 2;
    static double cx[4] /* was [2][2] */, dx, dy, deltay[21], dx1, dx2,
                  dxplus[2], dxmins[2], fcont, scrat[4] /* was [2][2] */,
                  p, r, t[21], x[2], y[21], sigma, rho, xpl[2], ypl[2],
                  xmin, ymin, xmax, ymax, a[42] /* was [21][2] */;
    static char capt[75], string[75];

/* identify program to user */
    printf("%s\n", "Program S5LSQ generates data");
    printf("%s\n", "corresponding to a Breit-Wigner law");
    printf("%s\n", "and fits a Breit-Wigner  OR a Gaussian.");
    printf("%s\n", "It shows fitted parameters with symmetric");
    printf("%s\n\n", "and asymmetric errors and confidence region.");
/* ask for input */
    printf("%s\n", "Enter size of measurement errors (> 0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - Fit to Breit-Wigner function");
    printf("%s\n", "2 - Fit to Gaussian");
    printf("%s", "> ");
    scanf("%li", &iswit);
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* generate data points corresponding to Breit-Wigner */
    x[0] = 0.;
    x[1] = 1.;
    rnstnr_(deltay, &c__21);
    for (i = 1; i <= 21; ++i) {
        t[i - 1] = (double) (i - 1) * .3 - 3.;
        y[i - 1] = bwfnct(x, &nr, &t[i - 1]);
        y[i - 1] += deltay[i - 1] * sigma;
        deltay[i - 1] = sigma;
    }
/* set first approximation of unknowns */
    x[0] = .5;
    x[1] = .5;
/* perform fit */
    nstep = 100;
    if (iswit == 1) {
        lsqnon_(bwfnct, t, y, deltay, &c__21, &nr, &nr, list, x,
               cx, &r, a, scrat, &nstep);
    } else {
        lsqnon_(gsfnct, t, y, deltay, &c__21, &nr, &nr, list, x,
               cx, &r, a, scrat, &nstep);
    }
    p = 1. - scchi2_(&r, &c__19);
    if (nstep < 0) {
        printf("%s%5li\n", "LSQNON ended with NSTEP = ", nstep);
    } else {
/* compute asymmetric errors */
        nstep = 100;
        if (iswit == 1) {
            lsqasn_(bwfnct, t, y, deltay, &c__21, &nr, &nr, list,
                   x, cx, &r, &c_d0, dxplus, dxmins, a, scrat, &nstep);
        } else {
            lsqasn_(gsfnct, t, y, deltay, &c__21, &nr, &nr, list,
                   x, cx, &r, &c_d0, dxplus, dxmins, a, scrat, &nstep);
        }
/* prepare graphic */
        sprintf(capt, "x_1#=%6.2lf, x_2#=%6.2lf, M=%6.2lf, P=%8.6lf",
                x[0], x[1], r, p);
        if (iswit == 1) {
            strncpy(capt + 46, " (fit to Breit-Wigner) ", 23);
        } else {
            strncpy(capt + 46, " (fit to Gaussian)     ", 23);
        }
        lcapt = strlen(capt);
        dx1 = sqrt(cx[0]);
        dx2 = sqrt(cx[3]);
        rho = cx[2] / (dx1 * dx2);
        xmin = x[0] - dx1 * 2.;
        ymin = x[1] - dx2 * 2.;
        xmax = x[0] + dx1 * 2.;
        ymax = x[1] + dx2 * 2.;
        gropen_();
        gropws_(&nws);
        grwncc_(&xmin, &xmax, &ymin, &ymax);
        grvwwc_(&c_dp2, &c_dp9, &c_dp17, &c_dp87);
        grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
        grfram_();
        grboun_();
        grtxtc_(&c_d1, capt, &lcapt);
        memset(string, ' ', 75);
        strncpy(string, "x_1", 3);
        grsclx_(string, &c__3);
        strncpy(string, "x_2", 3);
        grscly_(string, &c__3);
        grstcl_(&c__2);
        grdatp_(&c__1, &c_d1, x, &x[1], &dx1, &dx2, &rho);
        grstcl_(&c__3);
        for (i = 1; i <= 2; ++i) {
            if (i == 1) {
                xpl[0] = x[0] - dxmins[0];
            } else {
                xpl[0] = x[0] + dxplus[0];
            }
            xpl[1] = xpl[0];
            ypl[0] = ymin;
            ypl[1] = ymax;
            grplin_(&c__2, xpl, ypl);
        }
        for (i = 1; i <= 2; ++i) {
            if (i == 1) {
                ypl[0] = x[1] - dxmins[1];
            } else {
                ypl[0] = x[1] + dxplus[1];
            }
            ypl[1] = ypl[0];
            xpl[0] = xmin;
            xpl[1] = xmax;
            grplin_(&c__2, xpl, ypl);
        }
        fcont = r + 1.;
        grstcl_(&c__4);
        nx = 100;
        ny = 100;
        dx = (xmax - xmin) / nx;
        dy = (ymax - ymin) / ny;
        if (iswit == 1) {
            lsqcon_(&c__1, &c__2, &xmin, &ymin, &dx, &dy, &nx, &ny, &fcont, x,
                   t, y, deltay, &c__21, &nr, bwfnct);
        } else {
            lsqcon_(&c__1, &c__2, &xmin, &ymin, &dx, &dy, &nx, &ny, &fcont, x,
                   t, y, deltay, &c__21, &nr, gsfnct);
        }
        grclse_();
    }
    return 0;
} /* main */

/* ------------------------------------------------- */
double CBFUNC bwfnct(double const P_T *x, integer const P_T *nr,
                     double const P_T *t)
{
    /* System generated locals */
    double ret_val, d__1, d__2, d__3;

    /* Parameter adjustments */
    --x;

    /* Function Body */
/* Computing 2nd power */
    d__1 = x[2];
/* Computing 2nd power */
    d__2 = *t - x[1];
/* Computing 2nd power */
    d__3 = x[2];
    ret_val = d__1 * d__1 * 2. / (x[2] * 3.14159f * (d__2 * d__2 * 4. + d__3 *
             d__3));
    return ret_val;
} /* bwfnct */

/* ------------------------------------------------- */
double CBFUNC gsfnct(double const P_T *x, integer const P_T *nr,
                     double const P_T *t)
{
    /* System generated locals */
    double ret_val, d__1;

    /* Parameter adjustments */
    --x;

    /* Function Body */
/* Computing 2nd power */
    d__1 = (*t - x[1]) / x[2];
    ret_val = .39894228 / x[2] * exp(d__1 * d__1 * -.5);
    return ret_val;
} /* gsfnct */

File: runme/mcgen/local/datan/solsrc/s5lsq.f
      PROGRAM S5LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXNR=2,MAXN=21,N=21,T0=-3.D0,DELTAT=0.3D0)
      DIMENSION T(MAXN),DELTAY(MAXN),CX(MAXNR,MAXNR),A(MAXN,MAXNR)
      DIMENSION Y(MAXN),X(MAXNR),SCRAT(MAXNR,MAXNR)
      DIMENSION DXPLUS(MAXNR),DXMINS(MAXNR),XPL(2),YPL(2)
      CHARACTER*75 CAPT,STRING
      DIMENSION LIST(MAXNR)
      EXTERNAL BWFNCT,GSFNCT
      DOUBLE PRECISION BWFNCT,GSFNCT
      DATA NR /2/
C identify program to user
      WRITE(*,*)' Program S5LSQ generates data'
      WRITE(*,*)' corresponding to a Breit-Wigner law'
      WRITE(*,*)' and fits a Breit-Wigner  OR a Gaussian.'
      WRITE(*,*)' It shows fitted parameters with symmetric'
      WRITE(*,*)' and asymmetric errors and confidence region'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter size of measurement errors (> 0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - Fit to Breit-Wigner function'
      WRITE(*,*)' 2 - Fit to Gaussian'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C generate data points corresponding to Breit-Wigner
      X(1)=0.0D0
      X(2)=1.0D0
      CALL RNSTNR(DELTAY,N)
      DO 10 I=1,N
        T(I)=T0+DBLE(I-1)*DELTAT
        Y(I)=BWFNCT(X,NR,T(I))
        Y(I)=Y(I)+DELTAY(I)*SIGMA
        DELTAY(I)=SIGMA
10    CONTINUE
C set first approximation of unknowns
      X(1)=0.5D0
      X(2)=0.5D0
C perform fit
      NSTEP=100
      IF(ISWIT.EQ.1) THEN
        CALL LSQNON(BWFNCT,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,A,SCRAT,
     +  NSTEP)
      ELSE
        CALL LSQNON(GSFNCT,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,A,SCRAT,
     +  NSTEP)
      END IF
      P=1.D0-SCCHI2(R,N-NR)
      IF(NSTEP.LT.0) THEN
        WRITE(*,'(A,I5)')' LSQNON ended with NSTEP = ',NSTEP
      ELSE
C compute asymmetric errors
        NSTEP=100
        IF(ISWIT.EQ.1)THEN
          CALL LSQASN(BWFNCT,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,0.D0,
     +    DXPLUS,DXMINS,A,SCRAT,NSTEP)
        ELSE
          CALL LSQASN(GSFNCT,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,0.D0,
     +    DXPLUS,DXMINS,A,SCRAT,NSTEP)
        END IF
C prepare graphic
        CAPT='x_1#=******, x_2=#******, M=******, P=********'
        WRITE(CAPT(6:11),'(F6.2)')X(1)
        WRITE(CAPT(19:24),'(F6.2)')X(2)
        WRITE(CAPT(29:34),'(F6.2)')R
        WRITE(CAPT(39:46),'(F8.6)')P
        IF(ISWIT.EQ.1) THEN
          WRITE(CAPT(47:69),'(A)')' (fit to Breit-Wigner)'
          ELSE
          WRITE(CAPT(47:69),'(A)')' (fit to Gaussian)    '
        END IF
        LCAPT=69
        DX1=SQRT(CX(1,1))
        DX2=SQRT(CX(2,2))
        RHO=CX(1,2)/(DX1*DX2)
        XMIN=X(1)-2.D0*DX1
        YMIN=X(2)-2.D0*DX2
        XMAX=X(1)+2.D0*DX1
        YMAX=X(2)+2.D0*DX2
        CALL GROPEN
        CALL GROPWS(NWS)
        CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
        CALL GRVWWC(0.2D0,0.9D0,0.17D0,0.87D0)
        CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
        CALL GRFRAM
        CALL GRBOUN
        CALL GRTXTC(1.D0,CAPT,LCAPT)
        STRING='x_1'
        CALL GRSCLX(STRING,3)
        STRING='x_2'
        CALL GRSCLY(STRING,3)
        CALL GRSTCL(2)
        CALL GRDATP(1,1.D0,X(1),X(2),DX1,DX2,RHO)
        CALL GRSTCL(3)
        DO 20 I=1,2
          IF(I.EQ.1) THEN
            XPL(1)=X(1)-DXMINS(1)
          ELSE
            XPL(1)=X(1)+DXPLUS(1)
          END IF
          XPL(2)=XPL(1)
          YPL(1)=YMIN
          YPL(2)=YMAX
          CALL GRPLIN(2,XPL,YPL)
20      CONTINUE
        DO 30 I=1,2
          IF(I.EQ.1) THEN
            YPL(1)=X(2)-DXMINS(2)
          ELSE
            YPL(1)=X(2)+DXPLUS(2)
          END IF
          YPL(2)=YPL(1)
          XPL(1)=XMIN
          XPL(2)=XMAX
          CALL GRPLIN(2,XPL,YPL)
30      CONTINUE
        FCONT=R+1.D0
        CALL GRSTCL(4)
        NX=100
        NY=100
        DX=(XMAX-XMIN)/NX
        DY=(YMAX-YMIN)/NY
        IF(ISWIT.EQ.1) THEN
          CALL LSQCON(1,2,XMIN,YMIN,DX,DY,NX,NY,FCONT,X,T,Y,DELTAY,
     +    N,NR,BWFNCT)
        ELSE
          CALL LSQCON(1,2,XMIN,YMIN,DX,DY,NX,NY,FCONT,X,T,Y,DELTAY,
     +    N,NR,GSFNCT)
        END IF
        CALL GRCLSE
      END IF
      END
C-------------------------------------------------
      DOUBLE PRECISION FUNCTION BWFNCT(X,NR,T)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X(*)
      BWFNCT=2.D0*X(2)**2/((3.14159*X(2))*(4.D0*(T-X(1))**2+X(2)**2))
      END
C-------------------------------------------------
      DOUBLE PRECISION FUNCTION GSFNCT(X,NR,T)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X(*)
      GSFNCT=(.39894228D0/X(2))*EXP(-0.5D0*((T-X(1))/X(2))**2)
      END
